### PR TITLE
Sampling using General Basis Reduction

### PR DESCRIPTION
**Pull request by superty**
This patch adds a `Simplex` class. Currently this can consume a `FlatAffineConstraints`, and can find an integer sample point (if one exists) using general basis reduction. Additionally, it adds two classes `Matrix` and `Fraction`, which are used by `Simplex`.

This adds a `findSample` function to `FlatAffineStructures` which uses the above functionality of `Simplex`.

### CODE REVIEW

File: mlir/include/mlir/Analysis/AffineStructures.h [hunk 1]
```diff
@@ -139,6 +139,12 @@ class FlatAffineConstraints {
   // returns true, no integer solution to the equality constraints can exist.
   bool isEmptyByGCDTest() const;
 
+  // Find a sample point satisfying the constraints. This uses a branch and
+  // bound algorithm with general basis reduction.
+  //
+  // Returns such a point if one exists, or an empty llvm::Optional otherwise.
+  llvm::Optional<std::vector<int64_t>> findSample() const;
+
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/AffineStructures.h:151]: Would it make sense to add a method isIntegerEmpty() that uses first the fast isEmptyByGCD() test and falls back to GBR if that does not work?

File: mlir/include/mlir/Analysis/AffineStructures.h [hunk 2]
```diff
@@ -519,10 +525,13 @@ class FlatAffineConstraints {
...
 
-  /// A more expensive check to detect redundant inequalities thatn
-  /// removeTrivialRedundancy.
+  /// A more expensive check to detect redundant inequalities than
+  /// removeTrivialRedundancy. Uses Fourier-Motzkin.
   void removeRedundantInequalities();
 
+  /// A more expensive check to detect redundant constraints. Uses Simplex.
+  void removeRedundantConstraints();
+
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/AffineStructures.h]: I think this should also be dropped.

File: mlir/include/mlir/Analysis/AffineStructures.h [hunk 3]
```diff
@@ -519,10 +525,13 @@ class FlatAffineConstraints {
   /// constraints by their GCD and performs GCD tightening on inequalities.
   void removeTrivialRedundancy();
 
-  /// A more expensive check to detect redundant inequalities thatn
-  /// removeTrivialRedundancy.
+  /// A more expensive check to detect redundant inequalities than
+  /// removeTrivialRedundancy. Uses Fourier-Motzkin.
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/AffineStructures.h]: This is an unrelated change. Remove?

File: mlir/include/mlir/Analysis/Presburger/Fraction.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
+#ifndef MLIR_ANALYSIS_PRESBURGER_FRACTION_H
+#define MLIR_ANALYSIS_PRESBURGER_FRACTION_H
+
+namespace mlir {
+
+template <typename T>
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Fraction.h:19]: This could use a header comment.

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    UNMARK_EMPTY,
+  };
+
+  Simplex() = delete;
+  explicit Simplex(unsigned nVar);
+  explicit Simplex(const FlatAffineConstraints &constraints);
+  unsigned getNumRows() const;
+  unsigned getNumColumns() const;
+
+  /// \returns True is the tableau is empty (has conflicting constraints),
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: is -> if

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  /// False otherwise.
+  bool isEmpty() const;
+
+  /// Add an inequality to the tableau. If coeffs is c_0, c_1, ... c_n, where n
+  /// is the curent number of variables, then the corresponding inequality is
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: curent -> current

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  void addInequality(ArrayRef<int64_t> coeffs);
+
+  /// \returns the number of variables in the tableau.
+  unsigned numberVariables() const;
+
+  /// \returns the number of constraints in the tableau.
+  unsigned numberConstraints() const;
+
+  /// Add an equality to the tableau. If coeffs is c_0, c_1, ... c_n, where n
+  /// is the curent number of variables, then the corresponding inequality is
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: curent -> current

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  unsigned getSnapshot() const;
+
+  /// Rollback to a snapshot. This invalidates all later snapshots.
+  void rollback(unsigned snapshot);
+
+  /// Compute the maximum or minimum value of the given row, depending on
+  /// \p direction.
+  ///
+  /// \returns a {num, den} pair denoting the optimum, or a null value if no
+  /// optimum exists, i.e. if the expression is unbounded in this direction.
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: i.e.,

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+  llvm::Optional<Fraction<int64_t>> computeRowOptimum(Direction direction,
+                                                      unsigned row);
+
+  /// Compute the maximum or minimum value of the given expression, depending on
+  /// \p direction.
+  ///
+  /// \returns a {num, den} pair denoting the optimum, or a null value if no
+  /// optimum exists, i.e. if the expression is unbounded in this direction.
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: i.e.,

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+  /// Normalize the given row by removing common factors between the numerator
+  /// and the denominator.
+  void normalizeRow(unsigned row);
+
+  /// Swap the two rows in the tableau and associated data structures.
+  void swapRows(unsigned i, unsigned j);
+
+  /// Restore the unknown to a non-negative sample value.
+  ///
+  /// \returns True is the unknown was successfully restored to a non-negative
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: is -> if

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  void reduceBasis(Matrix<int64_t> &basis, unsigned level);
+
+  /// The number of rows in the tableau.
+  unsigned nRow;
+
+  /// The number of columns in the tableau, including the common denominator
+  /// and the constant column.
+  unsigned nCol;
+
+  /// The matrix represnting the tableau.
```
Comment (tobiasgrosser) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: represnting -> representing

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 12]
```diff
@@ -0,0 +1,929 @@
...
+      continue;
+    }
+
+    alpha.clear();
+    gbrSimplex.addEqualityForDirection(basis.getRow(i));
+    i++;
+  }
+}
+
+// Try to an integer sample point in the polytope.
```
Comment (tobiasgrosser) [inline mlir/lib/Analysis/Presburger/Simplex.cpp]: Try to FIND an integer sample point

**Review by tobiasgrosser — COMMENTED** [review]
Hey Arjun, some minor comments. Otherwise, this changeset looks great and a more in-depth review could take place in public. 

Do you have a plan for the commit message? In particular, one paragraph that motivates why you want to contribute this changeset.

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+template <typename INT>
+void Matrix<INT>::addToRow(unsigned sourceRow, unsigned targetRow, INT scale) {
+  if (scale == 0)
+    return;
+  for (unsigned col = 0; col < getNumColumns(); ++col)
+    data[targetRow][col] += scale * data[sourceRow][col];
+  return;
+}
+
+template <typename INT> void Matrix<INT>::dump() const {
```
Comment (Dinistro) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h]: Just a small remark from my side: In other MLIR classes I saw that they add a `print(os_stream &)` method as well. `dump` then just calls `print` with `llvm::errs()`.

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 14]
```diff
@@ -0,0 +1,929 @@
...
+  for (unsigned i = 0; i < constraints.getNumEqualities(); ++i)
+    addEquality(constraints.getEquality(i));
+}
+
+const Simplex::Unknown &Simplex::unknownFromIndex(int index) const {
+  assert(index != NULL_INDEX && "NULL_INDEX passed to unknownFromIndex");
+
+  if (index >= 0)
+    return var[index];
+  else
```
Comment (Dinistro) [inline mlir/lib/Analysis/Presburger/Simplex.cpp]: Just some nitpicking: According to the llvm coding style guide they prefer to omit an else after a return: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 15]
```diff
@@ -0,0 +1,929 @@
...
+  assert(row < nRow && "Invalid row");
+  return unknownFromIndex(rowVar[row]);
+}
+
+Simplex::Unknown &Simplex::unknownFromIndex(int index) {
+  assert(index != NULL_INDEX && "NULL_INDEX passed to unknownFromIndex");
+
+  if (index >= 0)
+    return var[index];
+  else
```
Comment (Dinistro) [inline mlir/lib/Analysis/Presburger/Simplex.cpp]: Same as above

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 16]
```diff
@@ -0,0 +1,929 @@
...
+  }
+  for (unsigned col = 0; col < nCol; ++col)
+    tableau(row, col) /= gcd;
+}
+
+bool Simplex::signMatchesDirection(int64_t elem, Direction direction) const {
+  assert(elem != 0 && "elem should not be 0");
+  if (direction == Direction::UP)
+    return elem > 0;
+  else
```
Comment (Dinistro) [inline mlir/lib/Analysis/Presburger/Simplex.cpp]: Same as above

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 17]
```diff
@@ -0,0 +1,929 @@
...
+  unsigned numCon = A.numberConstraints() + B.numberConstraints();
+  Simplex result(numVar);
+
+  result.tableau.resize(numCon, 2 + numVar);
+  result.empty = A.empty || B.empty;
+
+  auto indexFromBIndex = [&](int index) {
+    if (index < 0)
+      return ~(A.numberConstraints() + ~index);
+    else
```
Comment (Dinistro) [inline mlir/lib/Analysis/Presburger/Simplex.cpp]: Same as above

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 18]
```diff
@@ -0,0 +1,929 @@
...
+    auto width = simplex.computeRowOptimum(Simplex::Direction::UP, row);
+    assert(width && "Width should not be unbounded!");
+    dualDenom = simplex.tableau(row, 0);
+    dual.clear();
+    // The increment is i += 2 because equalities are added as two inequalities,
+    // one positive and one negative. We only want to process the positive ones.
+    for (unsigned i = simplexConstraintOffset; i < conIndex; i += 2) {
+      if (simplex.con[i].ownsRow)
+        dual.push_back(0);
+      else {
```
Comment (Dinistro) [inline mlir/lib/Analysis/Presburger/Simplex.cpp:603]: Here I would probably drop the brackets, but this is a discussion of taste I guess

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+  /// \returns the current sample point if it is integral. Otherwise, returns an
+  /// empty llvm::Optional.
+  llvm::Optional<std::vector<int64_t>> getSamplePointIfIntegral() const;
+
+  /// \returns an integral sample point if one exists, or an empty
+  /// llvm::Optional otherwise.
+  llvm::Optional<std::vector<int64_t>> findIntegerSample();
+
+  // Dump the tableau's internal state.
+  void dump() const;
```
Comment (Dinistro) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h:115]: Also here an additional print would be nice to have, but not necessary.

File: mlir/lib/Analysis/Presburger/Simplex.cpp (hunk 18 above)
Comment (superty) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:603]: Hm, unless this is generally followed in MLIR I prefer to keep the braces.

File: mlir/include/mlir/Analysis/Presburger/Fraction.h [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+int compare(const Fraction<T> &x, const Fraction<T> &y) {
+  auto diff = x.num * y.den - y.num * x.den;
+  if (diff > 0)
+    return +1;
+  if (diff < 0)
+    return -1;
+  return 0;
+}
+
+template <typename T>
```
Comment (chelini) [inline mlir/include/mlir/Analysis/Presburger/Fraction.h:44]: Why not: return (f.num >= 0) ? (fnum/f.den) : ((f.num / f.den) - (f.num % f.den != 0)); 
it saves 3 lines.

File: mlir/include/mlir/Analysis/Presburger/Fraction.h [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+T floor(Fraction<T> f) {
+  if (f.num >= 0)
+    return f.num / f.den;
+  else
+    return (f.num / f.den) - (f.num % f.den != 0);
+}
+
+template <typename T>
```
Comment (chelini) [inline mlir/include/mlir/Analysis/Presburger/Fraction.h:52]: same here.

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+/// The data is stored in the form of a vector of vectors.
+template <typename INT> class Matrix {
+public:
+  Matrix() = delete;
+
+  /// Construct a matrix with the specified number of rows and columns.
+  /// Initially, the values are default initialized.
+  Matrix(unsigned rows, unsigned columns);
+
+  Matrix(const Matrix &);
```
Comment (chelini) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h]: explicit.

File: mlir/include/mlir/Analysis/AffineStructures.h [hunk 23]
```diff
@@ -139,6 +139,16 @@ class FlatAffineConstraints {
   // returns true, no integer solution to the equality constraints can exist.
   bool isEmptyByGCDTest() const;
 
+  // \returns True if the set of constraints has no integer solution, and False
+  // if it does have a solution.
+  bool isIntegerEmpty() const;
+
+  // Find a sample point satisfying the constraints. This uses a branch and
```
Comment (chelini) [inline mlir/include/mlir/Analysis/AffineStructures.h:146]: why not /// instead of // ?

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <cassert>
+#include <vector>
+
+namespace mlir {
+
+/// This is a simple class to represent a resizable matrix.
+///
```
Comment (chelini) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h:25]: is this empty /// on purpose?

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+  for (unsigned row = 0; row < getNumRows(); ++row) {
+    os << row << " | ";
+    for (unsigned column = 0; column < getNumColumns(); ++column)
+      os << data[row][column] << " ";
+    os << '\n';
+  }
+}
+
+template <typename INT> void Matrix<INT>::dump() const {
+  print(llvm::errs());
```
Comment (chelini) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h:160]: why in llvm::errs and not llvm::outs?

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 26]
```diff
@@ -0,0 +1,919 @@
...
+    for (unsigned col = 1; col < nCol; ++col) {
+      if (col == pivotCol)
+        continue;
+      tableau(pivotRow, col) = -tableau(pivotRow, col);
+    }
+  }
+  normalizeRow(pivotRow);
+
+  for (unsigned row = 0; row < nRow; ++row) {
+    if (row == pivotRow)
```
Comment (chelini) [inline mlir/lib/Analysis/Presburger/Simplex.cpp:246]: can you or with the next condition? i.e., if ((row == pivotRow) || (tableau(/* */)) continue;

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 27]
```diff
@@ -0,0 +1,919 @@
...
+llvm::Optional<unsigned> Simplex::findPivotRow(llvm::Optional<unsigned> skipRow,
+                                               Direction direction,
+                                               unsigned col) const {
+  llvm::Optional<unsigned> retRow;
+  int64_t retElem, retConst;
+  for (unsigned row = 0; row < nRow; ++row) {
+    if (skipRow && row == *skipRow)
+      continue;
+    auto elem = tableau(row, col);
+    if (elem == 0)
```
Comment (chelini) [inline mlir/lib/Analysis/Presburger/Simplex.cpp:312]: Also here I would make an ||

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 28]
```diff
@@ -0,0 +1,919 @@
...
+  for (unsigned row = 0; row < nRow; ++row) {
+    for (unsigned col = 0; col < nCol; ++col)
+      os << tableau(row, col) << '\t';
+    os << '\n';
+  }
+  os << '\n';
+}
+
+void Simplex::dump() const {
+  print(llvm::errs());
```
Comment (chelini) [inline mlir/lib/Analysis/Presburger/Simplex.cpp:916]: outs() ?

File: mlir/include/mlir/Analysis/Presburger/Matrix.h (hunk 24 above)
Comment (superty) [reply mlir/include/mlir/Analysis/Presburger/Matrix.h:25]: Yes, I considered them to be different "paragraphs". Would it be better to put it in a single line?

File: mlir/include/mlir/Analysis/AffineStructures.h (hunk 23 above)
Comment (superty) [reply mlir/include/mlir/Analysis/AffineStructures.h:146]: For consistency with the rest of the file.

File: mlir/lib/Analysis/Presburger/Simplex.cpp (hunk 26 above)
Comment (superty) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:246]: Hm, I wrote it like this since that looks closer to the snippets in the coding standards: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

File: mlir/lib/Analysis/Presburger/Simplex.cpp (hunk 27 above)
Comment (superty) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:312]: See above.

File: mlir/include/mlir/Analysis/Presburger/Matrix.h (hunk 25 above)
Comment (superty) [reply mlir/include/mlir/Analysis/Presburger/Matrix.h:160]: FlatAffineConstraints dumps to llvm:errs(), so I did the same here.

File: mlir/lib/Analysis/Presburger/Simplex.cpp (hunk 28 above)
Comment (superty) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:916]: See above.

File: mlir/include/mlir/Analysis/Presburger/Matrix.h (hunk 22 above)
Comment (superty) [reply mlir/include/mlir/Analysis/Presburger/Matrix.h]: Removed, as it was unused.

File: mlir/include/mlir/Analysis/AffineStructures.h (hunk 23 above)
Comment (chelini) [reply mlir/include/mlir/Analysis/AffineStructures.h:146]: ok.

File: mlir/include/mlir/Analysis/Presburger/Matrix.h (hunk 24 above)
Comment (chelini) [reply mlir/include/mlir/Analysis/Presburger/Matrix.h:25]: No, it is ok.

File: mlir/include/mlir/Analysis/Presburger/Matrix.h (hunk 25 above)
Comment (chelini) [reply mlir/include/mlir/Analysis/Presburger/Matrix.h:160]: ok.

File: mlir/lib/Analysis/Presburger/Simplex.cpp (hunk 26 above)
Comment (chelini) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:246]: This is interesting. Not sure if this is followed in MLIR. Ignore my comment.

File: mlir/lib/Analysis/AffineStructures.cpp [hunk 29]
```diff
@@ -1035,6 +1036,20 @@ bool FlatAffineConstraints::isEmptyByGCDTest() const {
...
+// First, try the GCD test heuristic. If that doesn't find the set empty,
+// use the complete emptiness check provided by findIntegerSample, which gives
+// a definitive answer.
+bool FlatAffineConstraints::isIntegerEmpty() const {
+  if (isEmptyByGCDTest())
+    return true;
+  return !findIntegerSample().hasValue();
+}
+
+llvm::Optional<std::vector<int64_t>> FlatAffineConstraints::findIntegerSample() const {
```
Comment (Dinistro) [inline mlir/lib/Analysis/AffineStructures.cpp:1048]: MLIR normally uses `SmallVector` instead of `std::vector`. Perhaps this should be changed.

**Comment by Dinistro** [comment]
I think we cannot treat symbols the same way as dimensions in simplex. Unfortunately I have no idea how we could deal with symbols at all... 

See #8 for more information.
